assignment/impl/src/java/org/sakaiproject/assignment/impl/sort/AssignmentSubmissionComparator.java

```diff
@@ -62,11 +66,45 @@ public AssignmentSubmissionComparator(AssignmentService assignmentService, SiteS
 
     @Override
     public int compare(AssignmentSubmission a1, AssignmentSubmission a2) {
-        int result;
+        // Compare by UserSortNameComparator if appropriate
+        Optional<User> u1 = getIndividualSubmitterAsUser(a1);
+        Optional<User> u2 = getIndividualSubmitterAsUser(a2);
+        if (u1.isPresent() && u2.isPresent()) {
+            return userSortNameComparator.compare(u1.get(), u2.get());
+        }
+
         String name1 = getSubmitterSortname(a1);
         String name2 = getSubmitterSortname(a2);
-        result = compareString(name1, name2);
-        return result;
+        return compareString(name1, name2);
+    }
+
+    /**
+     * Gets the submitter of a submission for a non-group assignment.
+     * @return Optional<User> of the submission's submitter.
+     * Returns Optional.empty() if:
+     *     the submission is associated with a group assignment
+     *     the submission does not have exactly 1 associated submitter
+     *     the submitter cannot be retrieved via the UserDirectoryService
+     */
+    private Optional<User> getIndividualSubmitterAsUser(AssignmentSubmission submission) {
+        if (submission.getAssignment().getIsGroup()) {
+            return Optional.empty();
+        }
+
+        Set<AssignmentSubmissionSubmitter> submitters = submission.getSubmitters();
+        if (submitters.size() != 1) {
+            // On a non-group assignment, there should be precisely one submitter.
+            // When instructors submit on behalf of a user, it does not create a new submitter entry; rather it adds a submission property.
```

Comment:
hmm this comment doesn't seem correct. When a submission is submitted it should add all users in the group as submitters, but there won't be a submittee since it wasn't submitted by the student.

Reply:
Because of line 90, we're only concerned with assignments to be completed individually here, so groups shouldn't come into play. I know we used to be able to toggle between group vs individually completed assignments in... early Sakai 11? With dire consequences in doing so, of course. To my recollection, those cases were all cleaned out by the Sakai 12 assignment conversion job, so I think we're safe from group submissions showing up on an individually completed assignment at this point, right?

Reply:
So my code comment was interested in whether an instructor submitting on behalf of a student would add the instructor as an AssignmentSubmissionSubmitter on the submission, but in my testing that wasn't the case; they were added only as a submission property.

Reply:
correct and instructor (because of a permission) by default can't be a submitter.

Reply:
I'm still unclear - I'm testing on sakai master; my instructor has asn.submit, so to my understanding they have permission to be a submitter. When that instructor submits on behalf of a student, the AssignmentSubmissionSubmitter set still contains only 1 member: the student.

So I think my comment is still true, maybe I just need to elaborate:
```
/*
 *On a non-group assignment, there should be precisely one submitter.
 *Even when an instructor submits on behalf of a student, they do not create a new submitter entry (rather a submission property is created).
 * So expect the submitters set to contain only one user: the student. Log and abort otherwise.
 */
```
And I can move that comment block on top of:
`if (submitters.size() != 1) {`

Reply:
Updated accordingly